lib/routes/account.js

```diff
@@ -1745,18 +1781,68 @@ module.exports = (
 
         db.account(uid)
           .then(
-            (account) => {
+            (accountResult) => {
+              account = accountResult
+
+              isAccountVerification = butil.buffersAreEqual(code, account.emailCode)
+
               // This endpoint is not authenticated, so we need to look up
               // the target email address before we can check it with customs.
               return customs.check(request, account.email, 'recoveryEmailVerifyCode')
-                .then(
-                  () => account
-                )
             }
           )
+          .then(() => {
+
+            if (! isAccountVerification) {
+              // Check to see if this code belongs to an email in the emails table,
+              // at this point the code can only be a sign-in confirmation code
+              // or email code.
+              let verifiedEmail
+              return db.accountEmails(account.uid)
+                .then((emails) => {
+                  isEmailVerification = emails.some((email) => {
+                    if (email.emailCode && (code.toString('hex') === email.emailCode)) {
+                      verifiedEmail = email.email
+                      log.info({
+                        op: 'account.verifyEmail.secondary.started',
+                        uid: uidHex,
+                        code: request.payload.code
+                      })
+                      return true
+                    }
+                  })
+                })
+                .then(() => {
+                  // Verify email and send post verification email to the primary account
+                  if (isEmailVerification) {
+                    return db.verifyEmail(account, code)
+                      .then(function () {
+                        log.info({
+                          op: 'account.verifyEmail.secondary.confirmed',
+                          uid: uidHex,
+                          code: request.payload.code
+                        })
+                        return mailer.sendPostVerifySecondaryEmail(
+                          [],
+                          account,
+                          {
+                            acceptLanguage: request.app.acceptLanguage,
+                            secondaryEmail: verifiedEmail
+                          })
+                      })
+                  }
+                })
+            }
+
+            return true
+          })
           .then(
-            (account) => {
-              const isAccountVerification = butil.buffersAreEqual(code, account.emailCode)
+            () => {
+
+              if (isEmailVerification) {
```

Comment:
nit: should we add a doc here  (for future readers) about why we exit out early?